Task1.GenericDiagonalMatrix/Entities/DiagonalMatrix.cs

```diff
@@ -0,0 +1,74 @@
+﻿using System.Text;
+
+namespace Task1.GenericDiagonalMatrix
+{
+    //Followed docs.ms Guidlines @
+    //https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/events/how-to-publish-events-that-conform-to-net-framework-guidelines
+    internal class DiagonalMatrix<T>
+    {
+        public int Size => _size;
+        public event EventHandler<ElementChangedEventArgs<T>> ElementChanged;
+        private T?[] MatrixArray;
+        private readonly int _size;
+        public T? this[int i, int j]
+        {
+            get
+            {
+                if (i < 0 && j < 0 || i >= _size && j >= _size)
+                {
+                    throw new IndexOutOfRangeException();
+                }
+                if (i != j)
+                {
+                    return default(T);
+                }
+                else if (i == j)
+                {
+                    return MatrixArray[i];
+                }
+                else return default(T);
+            }
+            set
+            {
+                if (i >= 0 && j >= 0 || i <= _size && j <= _size || i == j)
```

Comment:
for setter IndexOutOfRangeException should be thrown under the same conditions as for getter

Reply:
In the set{  } ?
if (i < 0 && j < 0 || i >= _size && j >= _size)
   {
       throw new IndexOutOfRangeException();
   }